src/r1cs/verifier.rs

```diff
@@ -47,9 +36,20 @@ pub struct VerifierCS<'a, 'b> {
     /// variable assignments.
     num_vars: usize,
     V: Vec<CompressedRistretto>,
+
+    /// This list holds closures that will be called in the second phase of the protocol,
+    /// when non-randomized variables are committed.
+    /// After that, the option will flip to None and additional calls to `randomize_constraints`
+    /// will invoke closures immediately.
```

Comment:
This behavior (i.e., that calls to `randomize_constraints` invoke closures immediately in phase 2) seems important to mention in the user-facing docs, because it's what allows multiple challenges to be used for a set of constraints. I don't know how much we want to emphasize it, since doing something like
```
cs.specify_randomized_constraints(|cs, z| {
    cs.specify_randomized_constraints(|cs, w| {
        /* z, w = z, w */
    }
}
```
seems worse than 
```
cs.specify_randomized_constraints(|cs, z| {
    /* z, w = z, z*z */
}
```

Reply:
`z, w = z, z*z` is not safe, because if you need to generate powers of `z` and `w` down the road, those powers can overlap (see the use of `y` and `z` in the core bulletproofs protocol).

Reply:
Alternative, is to provide a different (sub)trait for CS which has method `challenge_scalar(label)->Scalar` which the user can call themselves.

Reply:
So it'll look something like this:

```
cs.specify_randomized_constraints(|cs| {
    let x = cs.challenge_scalar("x");
    let y = cs.challenge_scalar("y");
    ...
}
```

Reply:
I wanted to try an associated type `type RandomizedCS` set to Self in Prover/Verifier. I can check if it needs to be in scope.

Reply:
it might not only be cleaner in code, but also conceptually clean: the closure is actually specifying a _phase_, in which one can sample one or more challenges, so doing both of these things as separate calls matches this closer.

Reply:
Oh, a complication with doing this is that right now we keep the phase1/phase2 state internal to the CS impl (specifically the last function where it's consumed).  But if we want to have the `cs` parameter to the constraint callback implement a `Phase2CS` trait, it needs to be a different type so that the trait can be available only to that parameter and not on the original constraint system.

Reply:
Did some code changes: turns out, the user does not need to bring in scope RandomizedConstraintSystem to use challenges. And to your point, if the same type implements both traits, then user can define gadget over `<CS:RandomizedConstraintSystem>` which would be compatible with sending Prover/Verifier there (although it can't be used from within a correctly parametrized gadget in first-phase)?

Reply:
We can easily add another type by wrapping `Prover` in `RandomizingProver` and unwrapping after we are done with second phase.

Reply:
This would also make `mem::replace` and `Option<Vec>` go away.

Reply:
I think that the same type should *not* implement both traits, because it breaks the guarantees we achieve with this API (e.g., being able to allocate HL variables while forming constraints with no problems)...

Reply:
yeah, that's what i meant above - having two separate types for pre-randomization and randomization phases.

Reply:
Here's a sketch of a design that I think would work: we add a trait
```
pub trait Phase2CS: ConstraintSystem {
    /// Request a challenge scalar with the given domain separation label                                                                                                                                          
    fn challenge_scalar(&mut self, label: &'static [u8]) -> Scalar;
}
```
and have the callback take `cs: Phase2CS`.  Then we create (non-public) types `Phase2Prover` and `Phase2Verifier`, with `From` impls from `Prover` and `Verifier`.  The proving / verification logic moves into the `Phase2` types. When a user calls `Prover::prove()`, the method consumes the `self: Prover`, moves the data into a `Phase2Prover`, which executes the callbacks, passing itself (as a `Phase2CS` impl) to the callbacks, producing a proof (or verification) and then immediately returning the `Result`.

Reply:
yeah, i'm half-way doing that, but these types should be made public (although with private field, so only our crate can instantiate them).

Reply:
WIP here: https://github.com/dalek-cryptography/bulletproofs/compare/oleg/two-phase-impl...oleg/two-phase-impl-two-traits?expand=1

Reply:
Why should they be public?

Reply:
Looks like, they need to be public so that compiler can monomorphize the closures. They are anonymous types, but they need to be fully known nonetheless.

The good news is: the user does not have to even bring those in scope, and they cannot do anything with them outside the context of that `specify_randomized_constraints` call.